### PR TITLE
OP-1199 - Voucher Scheme new features (Subservices and subitems)

### PR DESCRIPTION
**Pull request by kevel-dev**

### CODE REVIEW

**Review by delcroip — APPROVED** [review]

**Comment by delcroip** [comment]
can you please fix the TC 
[test_mutation_update_service](https://github.com/openimis/openimis-be-medical_py/blob/62831402646f530d69f3c695ee10993769ef6fcd/medical/tests_api.py#L358)
[test_no_right_services_query](https://github.com/openimis/openimis-be-medical_py/blob/62831402646f530d69f3c695ee10993769ef6fcd/medical/tests_api.py#L184)

**Comment by delcroip** [comment]
And add test case for your features to make sure no one will break them

**Comment by delcroip** [comment]
poke @mngoe

**Comment by kevel-dev** [comment]
@delcroip i have added some test cases, all seem ok now

File: medical/gql_mutations.py [hunk 1]
```diff
@@ -110,13 +103,17 @@ def reset_item_or_service_before_update(item_service):
...
+        "packagetype", #service only
+        "manualPrice", #service only
     ]
     for field in fields:
         if hasattr(item_service, field):
             setattr(item_service, field, None)
 
 
 def update_or_create_item_or_service(data, user, item_service_model):
+    items = data.pop('items') if 'items' in data else []
```
Comment (delcroip) [inline medical/gql_mutations.py:115]: I need to check here, for the product this code lead to removal of all item/services if the item and services were not loaded.
I aim to fix  the product by having the default as None and in that case I could assume that no change was requested, this approach have the limitation that the FE should send the section items/services only when loaded beforehand; (FE part not implemented in Product therefore I consider [] as no change)

Comment (delcroip) [reply medical/gql_mutations.py:115]: this is not blocking for the merge

File: medical/gql_mutations.py [hunk 2]
```diff
@@ -91,7 +80,11 @@ class ItemOrServiceInputType(OpenIMISMutation.Input):
 
 class ServiceInputType(ItemOrServiceInputType):
     level = graphene.String(required=True)
+    packagetype = graphene.String(required=True)
```
Comment (delcroip) [inline medical/gql_mutations.py:83]: Having True create  regression issue, I would rather have a default value

```suggestion
    packagetype = graphene.String(required=False)
```

File: medical/gql_mutations.py [hunk 3]
```diff
@@ -131,12 +128,54 @@ def update_or_create_item_or_service(data, user, item_service_model):
         check_if_code_already_exists(data, item_service_model)
 
     if item_service_uuid:
+        item_service = item_service_model.objects.get(uuid=item_service_uuid)
+        # Delete Service present in the Database and absent in the list sent by FE
+        # Means that user click on delete button and old Service is not sent
+        serviceExisting = list()
+        serviceSent = list()
+        for ServiceList in ServiceService.objects.filter(servicelinkedService=item_service.id).all() :
```
Comment (delcroip) [inline medical/gql_mutations.py:136]: no validity check here
```suggestion
        for ServiceList in ServiceService.objects.filter(*filter_validity(),servicelinkedService=item_service.id).all() :
```

File: medical/gql_mutations.py [hunk 4]
```diff
@@ -131,12 +128,54 @@ def update_or_create_item_or_service(data, user, item_service_model):
...
+        for serviceToDeleteId in serviceToDelete:
+            ServiceService.objects.filter(
+                id=serviceToDeleteId,
+            ).delete()
+
+        # Delete Item present in the Database and absent in the list sent by FE
+        # Means that user click on delete button and old Ites is not sent
+        itemExisting = list()
+        itemSent = list()
+        for ItemList in ServiceItem.objects.filter(servicelinkedItem=item_service.id).all() :
```
Comment (delcroip) [inline medical/gql_mutations.py:152]: same as service
```suggestion
        for ItemList in ServiceItem.objects.filter(*filter_validity(), servicelinkedItem=item_service.id).all() :
```

File: medical/gql_mutations.py [hunk 5]
```diff
@@ -4,16 +4,18 @@
 
 import django.db.models.base
 import graphene
+from graphene import InputObjectType
 from core import assert_string_length, PATIENT_CATEGORY_MASK_ADULT, PATIENT_CATEGORY_MASK_MALE, \
     PATIENT_CATEGORY_MASK_MINOR, PATIENT_CATEGORY_MASK_FEMALE
```
Comment (delcroip) [inline medical/gql_mutations.py:9]: import filter validity
```suggestion
from core import (
    assert_string_length, 
    PATIENT_CATEGORY_MASK_ADULT, 
    PATIENT_CATEGORY_MASK_MALE, 
    PATIENT_CATEGORY_MASK_MINOR, 
    PATIENT_CATEGORY_MASK_FEMALE,
    filter_validity,
```

File: medical/migrations/.DS_Store
Comment (delcroip) [inline medical/migrations/.DS_Store:1]: why this file required ?

File: medical/migrations/0008_auto_20220804_1112.py
Comment (delcroip) [inline medical/migrations/0008_auto_20220804_1112.py:1]: need to be regenerated after model update

File: medical/models.py [hunk 6]
```diff
@@ -164,12 +166,14 @@ class Service(VersionedModel, ItemOrService):
     code = models.CharField(db_column='ServCode', max_length=6)
     name = models.CharField(db_column='ServName', max_length=100)
     type = models.CharField(db_column='ServType', max_length=1)
+    packagetype = models.CharField(db_column='ServPackageType', max_length=1, default="S")
```
Comment (delcroip) [inline medical/models.py:169]: FYI, a textchoice class will be more restrictive therefore avoid errors

here an example https://github.com/openimis/openimis-be-payroll_py/blob/3a862282e081356c1fa03a6868c606c294855894/payroll/models.py#L10

File: medical/models.py [hunk 7]
```diff
@@ -164,12 +166,14 @@ class Service(VersionedModel, ItemOrService):
...
+    manualPrice = models.BooleanField(default=False)
     level = models.CharField(db_column='ServLevel', max_length=1)
     price = models.DecimalField(db_column='ServPrice', max_digits=18, decimal_places=2)
     maximum_amount = models.DecimalField(db_column='MaximumAmount', max_digits=18, decimal_places=2, blank=True, null=True)
     care_type = models.CharField(db_column='ServCareType', max_length=1)
     frequency = models.SmallIntegerField(db_column='ServFrequency', blank=True, null=True)
-    patient_category = models.SmallIntegerField(db_column='ServPatCat')
+    patient_category = models.SmallIntegerField(db_column='ServPatCat', default="15")
```
Comment (delcroip) [inline medical/models.py:176]: Default should have a CONSTANT to make the code more readable

**Review by delcroip — CHANGES_REQUESTED** [review]
please check my comments, I don't see anything blocking but the required GQL element,

File: medical/models.py [hunk 8]
```diff
@@ -278,7 +282,47 @@ def save_history_on_update(sender, instance, **kwargs):
         instance.validity_from = now
 
 
-class ItemMutation(core_models.UUIDModel, ObjectMutation):
+class ServiceService(models.Model):
```
Comment (delcroip) [inline medical/models.py:285]: Should use businessHistoryModel instead of Model to have auditing and timecoorelation

**Comment by delcroip** [comment]
@kevel-dev @mngoe , in nepal we have a status that can be set to dead,  there is a reason that use an external table InsureeStatusReason so somehow it replace your changes

File: medical/migrations/.DS_Store
Comment (mngoe) [reply medical/migrations/.DS_Store:1]: it's not. Just temporary files from Mac OS not present in the gitignore, sorry.
@kevel-dev Please remove it

File: medical/gql_mutations.py (hunk 1 above)
Comment (kevel-dev) [reply medical/gql_mutations.py]: Fixed

File: medical/migrations/.DS_Store
Comment (kevel-dev) [reply medical/migrations/.DS_Store:1]: Fixed

File: medical/models.py (hunk 7 above)
Comment (kevel-dev) [reply medical/models.py]: Fixed

**Review by delcroip — APPROVED** [review]

**Pull request merged**